src/Futhark/Analysis/Interference.hs

```diff
@@ -0,0 +1,333 @@
+  analyseBody lumap inuse body
+
+-- | Perform interference analysis on the given statements. The result is a
+-- triple of the names currently in use, names that hit their last use somewhere
+-- within, and the resulting graph.
+analyseKernels ::
+  LocalScope KernelsMem m =>
+  LastUseMap ->
+  Stms KernelsMem ->
+  m (Graph VName)
+analyseKernels lumap stms = do
+  (_, _, graph) <- analyseKernels' lumap stms
+  -- Now, we need to insert edges between memory blocks which differ in size, if
+  -- they are in DefaultSpace.
+  spaces <- Map.filter (== DefaultSpace) <$> memSpaces stms
+  inv_size_map <-
+    memSizes stms
+      <&> flip Map.restrictKeys (Set.fromList $ Map.keys spaces)
+      <&> invertMap
+  let new_edges =
+        cartesian
+          (\x y -> if x /= y then cartesian makeEdge x y else mempty)
+          inv_size_map
+          inv_size_map
+  return $ graph <> new_edges
+
+-- | Return a mapping from memory blocks to their element sizes in the given
+-- statements.
+memSizes :: LocalScope KernelsMem m => Stms KernelsMem -> m (Map VName Int)
+memSizes stms =
+  inScopeOf stms $ fmap mconcat <$> mapM memSizesStm $ stmsToList stms
+  where
+    memSizesStm :: LocalScope KernelsMem m => Stm KernelsMem -> m (Map VName Int)
+    memSizesStm (Let pat _ e) = do
+      arraySizes <- fmap mconcat <$> mapM memElemSize $ patternNames pat
+      arraySizes' <- memSizesExp e
+      return $ arraySizes <> arraySizes'
+    memSizesExp :: LocalScope KernelsMem m => Exp KernelsMem -> m (Map VName Int)
+    memSizesExp (Op (Inner (SegOp segop))) =
+      let body = segopBody segop
+       in inScopeOf (kernelBodyStms body) $
+            fmap mconcat
+              <$> mapM memSizesStm
+              $ stmsToList $ kernelBodyStms body
+    memSizesExp (If _ then_body else_body _) = do
+      then_res <- memSizes $ bodyStms then_body
+      else_res <- memSizes $ bodyStms else_body
+      return $ then_res <> else_res
+    memSizesExp (DoLoop _ _ _ body) =
+      memSizes $ bodyStms body
+    memSizesExp _ = return mempty
+
+-- | Return a mapping from memory blocks to the space they are allocated in.
+memSpaces :: LocalScope KernelsMem m => Stms KernelsMem -> m (Map VName Space)
+memSpaces stms =
+  return $ foldMap getSpacesStm stms
+  where
+    getSpacesStm :: Stm KernelsMem -> Map VName Space
+    getSpacesStm (Let (Pattern [] [PatElem name _]) _ (Op (Alloc _ sp))) =
+      Map.singleton name sp
+    getSpacesStm (Let _ _ (Op (Alloc _ _))) = error "impossible"
+    getSpacesStm (Let _ _ (Op (Inner (SegOp segop)))) =
+      foldMap getSpacesStm $ kernelBodyStms $ segopBody segop
+    getSpacesStm (Let _ _ (If _ then_body else_body _)) =
+      foldMap getSpacesStm (bodyStms then_body)
+        <> foldMap getSpacesStm (bodyStms else_body)
+    getSpacesStm (Let _ _ (DoLoop _ _ _ body)) =
+      foldMap getSpacesStm (bodyStms body)
+    getSpacesStm _ = mempty
+
+segopBody :: SegOp lvl lore -> KernelBody lore
+segopBody segop =
```

Comment:
Maybe we should migrate this to the SegOp module; I think we are doing similar things elsewhere.